src/core/server/saved_objects/permission_control/acl.test.ts

```diff
@@ -0,0 +1,348 @@
+
+describe('acl', () => {
+  it('test has permission', () => {
+    const principals: Principals = {
+      users: ['user1'],
+      groups: [],
+    };
+    const permissions: Permissions = {
+      read: principals,
+    };
+    const acl = new ACL(permissions);
+    expect(
+      acl.hasPermission(['read'], {
+        users: ['user1'],
+        groups: [],
+      })
+    ).toEqual(true);
+
+    expect(
+      acl.hasPermission(['read'], {
+        users: ['user2'],
+        groups: [],
+      })
+    ).toEqual(false);
+
+    expect(
+      acl.hasPermission([], {
+        users: ['user2'],
+        groups: [],
+      })
+    ).toEqual(false);
+
+    const nullValue: unknown = undefined;
+    expect(acl.hasPermission(['read'], nullValue as Principals)).toEqual(false);
+    expect(acl.hasPermission(['read'], {})).toEqual(false);
+
+    acl.resetPermissions();
+    expect(acl.hasPermission(['read'], nullValue as Principals)).toEqual(false);
+    expect(acl.hasPermission(['read'], {})).toEqual(false);
+    expect(acl.hasPermission(['read'], principals)).toEqual(false);
+  });
+
+  it('test add permission', () => {
+    const acl = new ACL();
+    let result = acl
+      .addPermission(['read'], {
+        users: ['user1'],
+        groups: [],
+      })
+      .getPermissions();
+    expect(result?.read?.users).toEqual(['user1']);
+
+    acl.resetPermissions();
+    result = acl
+      .addPermission(['write', 'management'], {
+        users: ['user2'],
+        groups: ['group1', 'group2'],
+      })
+      .getPermissions();
+    expect(result?.write?.users).toEqual(['user2']);
+    expect(result?.management?.groups).toEqual(['group1', 'group2']);
+
+    acl.resetPermissions();
+    result = acl
+      .addPermission(['write', 'management'], {
+        users: ['user2'],
+      })
+      .addPermission(['write', 'management'], {
+        groups: ['group1'],
+      })
+      .getPermissions();
+    expect(result?.write?.users).toEqual(['user2']);
+    expect(result?.write?.groups).toEqual(['group1']);
+    expect(result?.management?.users).toEqual(['user2']);
+    expect(result?.management?.groups).toEqual(['group1']);
+
+    acl.resetPermissions();
+    const nullValue: unknown = undefined;
+    result = acl.addPermission([], nullValue as Principals).getPermissions();
+    expect(result).toEqual({});
+
+    acl.resetPermissions();
+    result = acl.addPermission(nullValue as string[], {} as Principals).getPermissions();
+    expect(result).toEqual({});
+  });
+
+  it('test remove permission', () => {
+    let principals: Principals = {
+      users: ['user1'],
+      groups: ['group1', 'group2'],
+    };
+    let permissions = {
+      read: principals,
+      write: principals,
+    };
+    let acl = new ACL(permissions);
+    let result = acl
+      .removePermission(['read'], {
+        users: ['user1'],
+      })
+      .removePermission(['write'], {
+        groups: ['group2'],
+      })
+      .removePermission(['write'], {
```

Comment:
one question about `removePermission`:

From the documentation, it seems like write permission includes the read permission. 
For example, if a user have both read and write permission to the saved object, and we call `removePermission` to remove the read permission. In this case, is the write permission still valid?

Or should we set a rule that we can not remove read permission while there is still a valid write permission?